cato/client.go

```diff
@@ -0,0 +1,332 @@
+	ApiKey          string                  `json:"apikey" yaml:"apikey"`
+	AccountId       int                     `json:"accountid" yaml:"accountid"`
+}
+
+func (c *CatoConfig) Validate() error {
+	if err := c.ClientOptions.Validate(); err != nil {
+		return fmt.Errorf("client_options: %v", err)
+	}
+	if c.AccountId == 0 {
+		return errors.New("missing account id")
+	}
+	if c.ApiKey == "" {
+		return errors.New("missing api key")
+	}
+	return nil
+}
+
+type CatoAdapter struct {
+	conf         CatoConfig
+	wg           sync.WaitGroup
+	isRunning    uint32
+	mRunning     sync.RWMutex
+	uspClient    *uspclient.Client
+	writeTimeout time.Duration
+
+	chStopped chan struct{}
+	wgSenders sync.WaitGroup
+
+	ctx context.Context
+}
+
+func NewCatoAdapter(conf CatoConfig) (*CatoAdapter, chan struct{}, error) {
+	a := &CatoAdapter{
+		conf:      conf,
+		isRunning: 1,
+	}
+
+	if a.conf.WriteTimeoutSec == 0 {
+		a.conf.WriteTimeoutSec = defaultWriteTimeout
+	}
+	a.writeTimeout = time.Duration(a.conf.WriteTimeoutSec) * time.Second
+
+	var err error
+	a.uspClient, err = uspclient.NewClient(conf.ClientOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	a.chStopped = make(chan struct{})
+
+	a.wgSenders.Add(1)
+	if marker == "" {
+		if _, err := os.Stat(configFile); err == nil {
+			content, err := ioutil.ReadFile(configFile)
+			if err != nil {
+				a.conf.ClientOptions.DebugLog(fmt.Sprintf("Error reading config file: %v", err))
+			} else {
+				marker = strings.TrimSpace(string(content))
+			}
+		}
+	}
+
+	go a.handleEvent(marker, strconv.Itoa(a.conf.AccountId), a.conf.ApiKey)
+
+	go func() {
+		a.wgSenders.Wait()
+		close(a.chStopped)
+	}()
+
+	return a, a.chStopped, nil
+}
+
+func (a *CatoAdapter) Close() error {
+	a.conf.ClientOptions.DebugLog("closing")
+
+	a.mRunning.Lock()
+	a.isRunning = 0
+	a.mRunning.Unlock()
+
+	a.wg.Done()
+	a.wg.Wait()
+
+	_, err := a.uspClient.Close()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *CatoAdapter) convertStructToMap(obj interface{}) map[string]interface{} {
+	data, err := json.Marshal(obj)
+	if err != nil {
+		return nil
+	}
+
+	var mapRepresentation map[string]interface{}
+	err = json.Unmarshal(data, &mapRepresentation)
+	if err != nil {
+		a.conf.ClientOptions.DebugLog(err.Error())
+		return nil
+	}
+
+	return mapRepresentation
+}
+
+func (a *CatoAdapter) handleEvent(marker string, account_id string, api_key string) uintptr {
+
+	start = time.Now()
+
+	eventFilterString := ""
+	eventSubfilterString := ""
+
+	iteration := 1
+	totalCount := 0
+
+	for {
+		query := fmt.Sprintf(`{
+  eventsFeed(accountIDs:[%q]
+    marker:%q
+    filters:[%s,%s])
+  {
+    marker
+    fetchedCount
+    accounts {
+      id
+      records {
+        time
+        fieldsMap
+      }
+    }
+  }
+}`, account_id, marker, eventFilterString, eventSubfilterString)
+
+		success, resp := a.send(query, account_id, api_key)
+		if !success {
+			a.conf.ClientOptions.DebugLog(fmt.Sprintf("%s", resp))
+			break
```

Comment:
One last tweak, since you are breaking out here, you will need to add a return value on line 232. Or can remove the uintptr return type from handleEvent